File: app/src/main/java/ua/fvadevand/screenlock/services/LockScreenAccessibilityService.java
package ua.fvadevand.screenlock.services;

import android.accessibilityservice.AccessibilityService;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ShortcutInfo;
import android.content.pm.ShortcutManager;
import android.graphics.drawable.Icon;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;
import android.view.accessibility.AccessibilityEvent;

import java.util.Collections;
import java.util.List;

import ua.fvadevand.screenlock.MainActivity;
import ua.fvadevand.screenlock.R;

public class LockScreenAccessibilityService extends AccessibilityService {

    public static final String ACTION_SCREEN_LOCK = "ACTION_SCREEN_LOCK";
    private static final String TAG = "LockScreenAccessibility";
    private static final String SHORTCUT_ID = "ua.fvadevand.screenlock_shortcut";
    private BroadcastReceiver mReceiver;

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {

    }

    @Override
    public void onInterrupt() {

    }

    @Override
    protected void onServiceConnected() {
        Log.i(TAG, "onServiceConnected: ");
        ShortcutManager shortcutManager = getSystemService(ShortcutManager.class);
        if (!shortcutManager.isRequestPinShortcutSupported()) return;
        List<ShortcutInfo> pinnedShortcuts = shortcutManager.getPinnedShortcuts();
        if (pinnedShortcuts.isEmpty()) {
            ShortcutInfo pinShortcutInfo = new ShortcutInfo.Builder(this, SHORTCUT_ID)
                    .setShortLabel(getString(R.string.label_short_shortcut))
                    .setLongLabel(getString(R.string.label_long_shortcut))
                    .setIcon(Icon.createWithResource(this, R.mipmap.ic_shortcut))
                    .setIntent(new Intent(this, MainActivity.class)
                            .setAction(ACTION_SCREEN_LOCK))
                    .build();
            shortcutManager.requestPinShortcut(pinShortcutInfo, null);
        } else {
            if (!pinnedShortcuts.get(0).isEnabled()) {
                shortcutManager.enableShortcuts(Collections.singletonList(SHORTCUT_ID));
            }
        }
        mReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (intent.getAction() != null && intent.getAction().equals(ACTION_SCREEN_LOCK)) {
                    performGlobalAction(GLOBAL_ACTION_LOCK_SCREEN);
                }
            }
        };
        LocalBroadcastManager.getInstance(this).registerReceiver(mReceiver, new IntentFilter(ACTION_SCREEN_LOCK));
    }

    @Override
    public boolean onUnbind(Intent intent) {
        ShortcutManager shortcutManager = getSystemService(ShortcutManager.class);
        if (shortcutManager.isRequestPinShortcutSupported()) {
            shortcutManager.disableShortcuts(Collections.singletonList(SHORTCUT_ID), getString(R.string.message_disabled_shortcut));
            LocalBroadcastManager.getInstance(this).unregisterReceiver(mReceiver);
        }
        return super.onUnbind(intent);
    }
}

File: app/src/main/java/ua/fvadevand/screenlock/services/LockScreenAccessibilityService.kt
package ua.fvadevand.screenlock.services

import android.accessibilityservice.AccessibilityService
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.graphics.drawable.Icon
import android.view.accessibility.AccessibilityEvent
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import ua.fvadevand.screenlock.R
import ua.fvadevand.screenlock.ui.MainActivity

class LockScreenAccessibilityService : AccessibilityService() {
    private val receiver: BroadcastReceiver by lazy {
        object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                if (intent.action != null && intent.action == ACTION_SCREEN_LOCK) {
                    performGlobalAction(GLOBAL_ACTION_LOCK_SCREEN)
                }
            }
        }
    }

    companion object {
        const val ACTION_SCREEN_LOCK = "ua.fvadevand.screenlock.action.SCREEN_LOCK"
        private const val TAG = "LockScreenAccessibility"
        private const val SHORTCUT_ID = "ua.fvadevand.screenlock.shortcut.SCREEN_LOCK"
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent) {

    }

    override fun onInterrupt() {

    }

    override fun onServiceConnected() {
        getSystemService(ShortcutManager::class.java)?.apply {
            if (!isRequestPinShortcutSupported) return
            if (pinnedShortcuts.isEmpty()) {
                val pinShortcutInfo =
                    ShortcutInfo.Builder(this@LockScreenAccessibilityService, SHORTCUT_ID)
                        .setShortLabel(getString(R.string.label_short_shortcut))
                        .setLongLabel(getString(R.string.label_long_shortcut))
                        .setIcon(
                            Icon.createWithResource(
                                this@LockScreenAccessibilityService,
                                R.mipmap.ic_shortcut
                            )
                        )
                        .setIntent(
                            Intent(this@LockScreenAccessibilityService, MainActivity::class.java)
                                .setAction(ACTION_SCREEN_LOCK)
                        )
                        .build()
                requestPinShortcut(pinShortcutInfo, null)
            } else if (!pinnedShortcuts[0].isEnabled) {
                enableShortcuts(listOf(SHORTCUT_ID))
            }
            LocalBroadcastManager.getInstance(this@LockScreenAccessibilityService)
                .registerReceiver(receiver, IntentFilter(ACTION_SCREEN_LOCK))
        }
    }

    override fun onUnbind(intent: Intent): Boolean {
        getSystemService(ShortcutManager::class.java)?.apply {
            if (isRequestPinShortcutSupported) {
                disableShortcuts(
                    listOf(SHORTCUT_ID),
                    getString(R.string.message_disabled_shortcut)
                )
                LocalBroadcastManager.getInstance(this@LockScreenAccessibilityService)
                    .unregisterReceiver(receiver)
            }
        }
        return super.onUnbind(intent)
    }
}
